### PR TITLE
Etcd bootstrapping

### PR DESCRIPTION
**Pull request by gaurav36**
etcd bootstrapping in glusterd 2.0
    
    Adding member upon peer probe in the etcd cluster.
    
    With this change it will add remote host url during peer probe of remote host
    in the cluster.
    
    Issue #72
    
    Signed-off-by: Gaurav Kumar Garg <garg.gaurav52@gmail.com>

### CODE REVIEW

**Comment by gaurav36** [comment]
This PR now depend on https://github.com/gluster/glusterd2/pull/76 . Once #76 will merge then need to do some modification in this PR.

File: etcdmgmt/etcd-mgmt.go [hunk 1]
```diff
@@ -15,31 +15,12 @@ import (
...
-
-	initialAdvPeerUrls := "http://" + hostname + ":2380"
 
-	etcdCmd := exec.Command(ExecName,
-		"-listen-client-urls", listenClientUrls,
-		"-advertise-client-urls", advClientUrls,
-		"-listen-peer-urls", listenPeerUrls,
-		"-initial-advertise-peer-urls", initialAdvPeerUrls,
-		"--initial-cluster", "default="+listenPeerUrls)
+	etcdCmd := exec.Command(ExecName, args...)
```
Comment (atinmu) [inline etcdmgmt/etcd-mgmt.go]: Why are you bringing up etcd with no args here?

File: etcdmgmt/etcd-mgmt.go [hunk 2]
```diff
@@ -79,3 +59,44 @@ func StartEtcd() (*exec.Cmd, error) {
...
+	listenPeerUrls := "http://" + hostname + ":2380"
+	initialAdvPeerUrls := "http://" + hostname + ":2380"
+
+	args := []string{"-listen-client-urls", listenClientUrls,
+		"-advertise-client-urls", advClientUrls,
+		"-listen-peer-urls", listenPeerUrls,
+		"-initial-advertise-peer-urls", initialAdvPeerUrls}
+
+	return StartEtcd(args)
+}
```
Comment (atinmu) [inline etcdmgmt/etcd-mgmt.go]: Looking at these three flavours I firmly believe the functions can be still made one with varying args.

File: rpc/client/peer-rpc-clnt.go [hunk 3]
```diff
@@ -49,3 +49,34 @@ func ValidateAddPeer(p *peer.PeerAddRequest) (*services.RPCPeerAddResp, error) {
 	}
 	return rsp, nil
 }
+
+func ExportEtcdEnvVar(argv *services.RPCEtcdEnvReq) (*services.RPCEtcdEnvResp, error) {
```
Comment (atinmu) [inline rpc/client/peer-rpc-clnt.go]: Why do you need a different service, isn't it part of the peer add flow?

File: rpc/services/peer-rpc-svc.go [hunk 4]
```diff
@@ -39,3 +48,56 @@ func (p *PeerService) ValidateAdd(args *RPCPeerAddReq, reply *RPCPeerAddResp) er
 
 	return nil
 }
+
+func storeEtcdEnv(env *RPCEtcdEnvReq) error {
```
Comment (atinmu) [inline rpc/services/peer-rpc-svc.go]: Is this a function which is for services? I think this can definitely move to utils.

File: rpc/services/peer-rpc-svc.go [hunk 5]
```diff
@@ -39,3 +48,56 @@ func (p *PeerService) ValidateAdd(args *RPCPeerAddReq, reply *RPCPeerAddResp) er
...
+	if err := ioutil.WriteFile(etcdConfFile, []byte("ETCD_INITIAL_CLUSTER="+*env.InitialCluster), os.ModePerm); err != nil {
+		return err
+	}
+	if err := ioutil.WriteFile(etcdConfFile, []byte("ETCD_INITIAL_CLUSTER_STATE"+*env.ClusterState), os.ModePerm); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (etcd *ExportEtcd) ExportAndStoreEtcdEnv(env *RPCEtcdEnvReq, reply *RPCEtcdEnvResp) error {
```
Comment (atinmu) [inline rpc/services/peer-rpc-svc.go]: Do you really need a different type 'ExportEtcd' ?

File: rpc/client/peer-rpc-clnt.go (hunk 3 above)
Comment (atinmu) [inline rpc/client/peer-rpc-clnt.go]: This actually should be Add() function as we'd need to communicate additional information related to etcd env variables but in the add flow.

**Comment by atinmu** [comment]
@gaurav36 : I've not looked at the patch in depth. You'd first need to fix the structuring part of it and then I'd look at it.

**Comment by gaurav36** [comment]
@atinmu  once https://github.com/gluster/glusterd2/pull/76 merged then it will be easier to add missing things and address more comment. some part of this patch depend on https://github.com/gluster/glusterd2/pull/76

File: etcdmgmt/etcd-mgmt.go (hunk 1 above)
Comment (gaurav36) [inline etcdmgmt/etcd-mgmt.go]: Assigning the pre determined argument code present in this file only.

File: etcdmgmt/etcd-mgmt.go (hunk 2 above)
Comment (gaurav36) [inline etcdmgmt/etcd-mgmt.go]: i will further check about it.

File: rpc/client/peer-rpc-clnt.go (hunk 3 above)
Comment (gaurav36) [inline rpc/client/peer-rpc-clnt.go]: calling of this missed in this patch. i have already mentioned  in my previous PR that i need to do some modification in this patch.

File: rpc/services/peer-rpc-svc.go (hunk 5 above)
Comment (gaurav36) [inline rpc/services/peer-rpc-svc.go]: can also be done with peer service.

**Comment by gaurav36** [comment]
WIP

File: commands/peers/addpeer.go [hunk 6]
```diff
@@ -43,11 +50,71 @@ func addPeerHandler(w http.ResponseWriter, r *http.Request) {
 		rest.SendHTTPError(w, http.StatusInternalServerError, *rsp.OpError)
 		return
 	}
+
+	h, _ := os.Hostname()
```
Comment (atinmu) [inline commands/peers/addpeer.go]: error shouldn't be overlooked here.

File: commands/peers/addpeer.go [hunk 7]
```diff
@@ -43,11 +50,71 @@ func addPeerHandler(w http.ResponseWriter, r *http.Request) {
...
+	c, err := etcdclient.New(etcdclient.Config{Endpoints: []string{"http://" + h + ":2379"}})
```
Comment (atinmu) [inline commands/peers/addpeer.go]: As @kshlm pointed out earlier that we'd need to store etcd client's globally and create it once at the life time of GlusterD such that we don't have to create it for each request.

Comment (gaurav36) [inline commands/peers/addpeer.go]: done in next updated PR

File: commands/peers/addpeer.go (hunk 6 above)
Comment (gaurav36) [inline commands/peers/addpeer.go]: done in next updated PR

**Comment by gaurav36** [comment]
IP address resolution need to be done. Currently this PR using os.Hostname for getting the hostname. Most of the time etcd failing resolving hostname.

File: context/context.go [hunk 8]
```diff
@@ -43,6 +45,22 @@ func initOpVersion() {
 	OpVersion = MaxOpVersion
 }
 
+func initEtcdClient() error {
+	h, err := os.Hostname()
+	if err != nil {
+		log.Fatal("Could not able to get hostname")
+		return err
```
Comment (atinmu) [inline context/context.go]: log.Fatal does a exit (), so return doesn't make sense here.

File: commands/peers/addpeer.go [hunk 9]
```diff
@@ -43,11 +50,65 @@ func addPeerHandler(w http.ResponseWriter, r *http.Request) {
 		rest.SendHTTPError(w, http.StatusInternalServerError, *rsp.OpError)
 		return
 	}
+
+	c := context.EtcdClient
+	// Add member to etcd server
+	mAPI := etcdclient.NewMembersAPI(c)
+	member, e := mAPI.Add(etcdcontext.Background(), "http://"+p.Name+":2380")
+	if e != nil {
+		log.WithField("err", e).Error("Failed to add member")
```
Comment (atinmu) [inline commands/peers/addpeer.go]: The error message could be made better here. Failed to add member into etcd cluster would make it more clear.

File: commands/peers/addpeer.go [hunk 10]
```diff
@@ -43,11 +50,65 @@ func addPeerHandler(w http.ResponseWriter, r *http.Request) {
...
+		rest.SendHTTPError(w, http.StatusInternalServerError, e.Error())
+		return
+	}
+
+	newID := member.ID
+	newName := "name_" + p.Name
```
Comment (atinmu) [inline commands/peers/addpeer.go]: A better name than "_name" required. Probably "ETCD_" would be a better fit here.

File: commands/peers/addpeer.go [hunk 11]
```diff
@@ -43,11 +50,65 @@ func addPeerHandler(w http.ResponseWriter, r *http.Request) {
...
+
+	log.WithFields(log.Fields{
+		"New member ": newName,
+		"member Id ":  newID,
+	}).Info("New member added to the cluster")
+
+	mlist, e := mAPI.List(etcdcontext.Background())
+	if e != nil {
+		log.WithField("err", e).Error("Failed to list member")
```
Comment (atinmu) [inline commands/peers/addpeer.go]: Failed to list etcd cluster would be a better log here.

File: commands/peers/addpeer.go [hunk 12]
```diff
@@ -43,11 +50,65 @@ func addPeerHandler(w http.ResponseWriter, r *http.Request) {
...
+	log.WithField("ETCD_INITIAL_CLUSTER", strings.Join(conf, ",")).Info("ETCD_INITIAL_CLUSTER")
+	log.Info("ETCD_INITIAL_CLUSTER_STATE\"existing\"")
+
+	var etcdEnv peer.PeerETCDEnv
+	etcdEnv.PeerName = p.Name
+	etcdEnv.Name = newName
+	etcdEnv.InitialCluster = strings.Join(conf, ",")
+	etcdEnv.ClusterState = "existing"
+
+	etcdrsp, e := client.AddEtcdEnvVar(&etcdEnv)
```
Comment (atinmu) [inline commands/peers/addpeer.go]: s/AddEtcdEnvVar/ConfigureETCDEnv

File: commands/peers/addpeer.go [hunk 13]
```diff
@@ -43,11 +50,65 @@ func addPeerHandler(w http.ResponseWriter, r *http.Request) {
...
+	if e != nil {
+		log.WithField("err", e).Error("Failed to set etcd env in remote node")
+		rest.SendHTTPError(w, http.StatusInternalServerError, *etcdrsp.OpError)
+		return
+	}
+
 	if e = peer.AddOrUpdatePeer(p); e != nil {
+		log.WithField("err", e).Error("Failed to add peer into the store")
```
Comment (atinmu) [inline commands/peers/addpeer.go]: Mentioning the peer name/address would be a must here in the log.

File: etcdmgmt/etcd-mgmt.go [hunk 14]
```diff
@@ -94,13 +81,13 @@ func StartETCD() (*os.Process, error) {
...
 
-	err = etcdCmd.Start()
+	err := etcdCmd.Start()
 	if err != nil {
 		log.WithField("error", err.Error()).Error("Could not start etcd daemon.")
 		return nil, err
 	}
 
-	if check := checkHealth(15, listenClientUrls); check != true {
+	if check := checkHealth(15, args[1]); check != true {
```
Comment (atinmu) [inline etcdmgmt/etcd-mgmt.go]: Can we add a comment here to mention why we are checking health here?

File: etcdmgmt/etcd-mgmt.go [hunk 15]
```diff
@@ -160,3 +147,115 @@ func isETCDStartNeeded() (bool, int) {
 	}
 	return start, pid
 }
+
```
Comment (atinmu) [inline etcdmgmt/etcd-mgmt.go]: Every function at the beginning should have a comment what that function does. Please change accordingly for all new functions

File: etcdmgmt/etcd-mgmt.go [hunk 16]
```diff
@@ -160,3 +147,115 @@ func isETCDStartNeeded() (bool, int) {
...
+func EtcdStartInit() (*os.Process, error) {
```
Comment (atinmu) [inline etcdmgmt/etcd-mgmt.go]: s/Etcd/ETCD, take it as a global comment

Comment (atinmu) [inline etcdmgmt/etcd-mgmt.go]: Can we rename this function to SpawnETCD() ?

File: etcdmgmt/etcd-mgmt.go [hunk 17]
```diff
@@ -160,3 +147,115 @@ func isETCDStartNeeded() (bool, int) {
...
+	// Check whether etcd environment variable present or not
+	// If it present then start etcd without --initial-cluster flag
+	// other wise start etcd normally.
+
+	file, err := os.Open(etcdConfFile)
+	if err != nil {
```
Comment (atinmu) [inline etcdmgmt/etcd-mgmt.go]: Shouldn't you need to consider that if the file doesn't exist then only you have to start etcd in standalone mode? What I mean here is os.Open() can fail because of many reasons, we'd need consider only the error when the file doesn't exist and then proceed for a standalone boot otherwise its a fatal case and etcd can't be started.

File: etcdmgmt/etcd-mgmt.go [hunk 18]
```diff
@@ -160,3 +147,115 @@ func isETCDStartNeeded() (bool, int) {
...
+		log.Info("Starting/Restarting etcd for a initial node")
+		return StartInitialEtcd()
```
Comment (atinmu) [inline etcdmgmt/etcd-mgmt.go]: s/StartInitialEtcd/StartStandAloneETCD ?

File: etcdmgmt/etcd-mgmt.go [hunk 19]
```diff
@@ -160,3 +147,115 @@ func isETCDStartNeeded() (bool, int) {
...
+		scanner := bufio.NewScanner(file)
+		scanner.Split(bufio.ScanLines)
+		for scanner.Scan() {
+			linestr := scanner.Text()
+			etcdenv := strings.Split(linestr, "=")
+
+			etcdEnvKey := etcdenv[0]
+			etcdEnvData := etcdenv[1]
+			envlen := len(etcdenv)
+			for i := 2; i < envlen; i++ {
```
Comment (atinmu) [inline etcdmgmt/etcd-mgmt.go]: Could you explain this for loop, why is this needed?

File: etcdmgmt/etcd-mgmt.go [hunk 20]
```diff
@@ -160,3 +147,115 @@ func isETCDStartNeeded() (bool, int) {
...
+			}
+
+			// setting etcd environment variable
+			os.Setenv(etcdEnvKey, etcdEnvData)
+		}
+
+		hostname, err := os.Hostname()
+		if err != nil {
+			log.Fatal("Could not able to get hostname")
+			return nil, err
```
Comment (atinmu) [inline etcdmgmt/etcd-mgmt.go]: return is not needed.

File: etcdmgmt/etcd-mgmt.go [hunk 21]
```diff
@@ -160,3 +147,115 @@ func isETCDStartNeeded() (bool, int) {
...
+
+		listenClientUrls := "http://" + hostname + ":2379"
+		advClientUrls := "http://" + hostname + ":2379"
+		listenPeerUrls := "http://" + hostname + ":2380"
+		initialAdvPeerUrls := "http://" + hostname + ":2380"
+
+		args := []string{"-listen-client-urls", listenClientUrls,
+			"-advertise-client-urls", advClientUrls,
+			"-listen-peer-urls", listenPeerUrls,
+			"-initial-advertise-peer-urls", initialAdvPeerUrls}
```
Comment (atinmu) [inline etcdmgmt/etcd-mgmt.go]: I can see this snippet of forming args is been used in multiple places, so definitely can be reusable with the help of a function.

File: etcdmgmt/etcd-mgmt.go [hunk 22]
```diff
@@ -160,3 +147,115 @@ func isETCDStartNeeded() (bool, int) {
...
+		"-advertise-client-urls", advClientUrls,
+		"-listen-peer-urls", listenPeerUrls,
+		"-initial-advertise-peer-urls", initialAdvPeerUrls,
+		"--initial-cluster", "default=" + listenPeerUrls}
+
+	return StartETCD(args)
+}
+
+func etcdStop(etcdctx *os.Process) error {
+	err := etcdctx.Kill()
```
Comment (atinmu) [inline etcdmgmt/etcd-mgmt.go]: s/etcdStop/StopETCD

File: etcdmgmt/etcd-mgmt.go [hunk 23]
```diff
@@ -160,3 +147,115 @@ func isETCDStartNeeded() (bool, int) {
...
+	args := []string{"-listen-client-urls", listenClientUrls,
+		"-advertise-client-urls", advClientUrls,
+		"-listen-peer-urls", listenPeerUrls,
+		"-initial-advertise-peer-urls", initialAdvPeerUrls,
+		"--initial-cluster", "default=" + listenPeerUrls}
+
+	return StartETCD(args)
+}
+
+func etcdStop(etcdctx *os.Process) error {
```
Comment (atinmu) [inline etcdmgmt/etcd-mgmt.go]: s/etcdctx/etcdCtx

File: etcdmgmt/etcd-mgmt.go [hunk 24]
```diff
@@ -160,3 +147,115 @@ func isETCDStartNeeded() (bool, int) {
...
+	err := etcdctx.Kill()
+	if err != nil {
+		log.Error("Could not able to kill etcd daemon")
+		return err
+	}
+	etcdctx.Wait()
```
Comment (atinmu) [inline etcdmgmt/etcd-mgmt.go]: error should be handled for Wait ()

File: peer/peer.go [hunk 25]
```diff
@@ -17,3 +17,12 @@ type PeerAddRequest struct {
 	Addresses []string `json:"addresses"`
 	Name      string   `json:"name,omitempty"`
 }
+
+// PeerETCDEnv represents the structure which holds the ETCD env variables to be
+// set at the remote peer
+type PeerETCDEnv struct {
```
Comment (atinmu) [inline peer/peer.go]: s/PeerETCDEnv/ETCDEnvConfig

File: rpc/services/peer-rpc-svc.go [hunk 26]
```diff
@@ -2,17 +2,24 @@ package services
...
 )
 
 type PeerService int
 
 var (
-	opRet   int32
-	opError string
+	opRet        int32
+	opError      string
+	etcdConfDir  = "/var/lib/glusterd/"
```
Comment (atinmu) [inline rpc/services/peer-rpc-svc.go]: etcd related configuration can be defined in config.go as well

File: rpc/services/peer-rpc-svc.go [hunk 27]
```diff
@@ -39,3 +46,63 @@ func (p *PeerService) ValidateAdd(args *RPCPeerAddReq, reply *RPCPeerAddResp) er
 
 	return nil
 }
+
+func storeEtcdEnv(env *RPCEtcdEnvReq) error {
+	utils.InitDir(etcdConfDir)
+	fp, err := os.OpenFile(etcdConfFile, os.O_CREATE|os.O_WRONLY, 0666)
+	if err != nil {
+		log.WithField("err", err).Error("Failed to open etcdConfFile")
```
Comment (atinmu) [inline rpc/services/peer-rpc-svc.go]: etcdConfFile should be logged here too

File: rpc/services/peer-rpc-svc.go [hunk 28]
```diff
@@ -39,3 +46,63 @@ func (p *PeerService) ValidateAdd(args *RPCPeerAddReq, reply *RPCPeerAddResp) er
...
+		return err
+	}
+	defer fp.Close()
+
+	if _, err = fp.WriteString("ETCD_NAME=" + *env.Name + "\n"); err != nil {
+		log.WithField("err", err).Error("Failed to write Environment variable to etcdConfFile")
```
Comment (atinmu) [inline rpc/services/peer-rpc-svc.go]: same as above

File: rpc/services/peer-rpc-svc.go [hunk 29]
```diff
@@ -39,3 +46,63 @@ func (p *PeerService) ValidateAdd(args *RPCPeerAddReq, reply *RPCPeerAddResp) er
...
+		return err
+	}
+
+	if _, err = fp.WriteString("ETCD_INITIAL_CLUSTER=" + *env.InitialCluster + "\n"); err != nil {
+		log.WithField("err", err).Error("Failed to write Environment variable to etcdConfFile")
```
Comment (atinmu) [inline rpc/services/peer-rpc-svc.go]: same as above

File: rpc/services/peer-rpc-svc.go [hunk 30]
```diff
@@ -39,3 +46,63 @@ func (p *PeerService) ValidateAdd(args *RPCPeerAddReq, reply *RPCPeerAddResp) er
...
+		return err
+	}
+
+	if _, err = fp.WriteString("ETCD_INITIAL_CLUSTER_STATE=" + *env.ClusterState + "\n"); err != nil {
+		log.WithField("err", err).Error("Failed to write Environment variable to etcdConfFile")
```
Comment (atinmu) [inline rpc/services/peer-rpc-svc.go]: same as above

File: rpc/services/peer-rpc-svc.go [hunk 31]
```diff
@@ -39,3 +46,63 @@ func (p *PeerService) ValidateAdd(args *RPCPeerAddReq, reply *RPCPeerAddResp) er
...
+func (etcd *PeerService) ExportAndStoreEtcdEnv(env *RPCEtcdEnvReq, reply *RPCEtcdEnvResp) error {
+	opRet = 0
+	opError = ""
+
+	// Exporting etcd environment variable
+	os.Setenv("ETCD_NAME", *env.Name)
+	os.Setenv("ETCD_INITIAL_CLUSTER", *env.InitialCluster)
+	os.Setenv("ETCD_INITIAL_CLUSTER_STATE", *env.ClusterState)
+
+	// Storing there envioronment variable locally. So that upon glusterd
```
Comment (atinmu) [inline rpc/services/peer-rpc-svc.go]: s/there/the

File: rpc/services/peer-rpc-svc.go [hunk 32]
```diff
@@ -39,3 +46,63 @@ func (p *PeerService) ValidateAdd(args *RPCPeerAddReq, reply *RPCPeerAddResp) er
...
+	// restart we can set these environment variable again
+	err := storeEtcdEnv(env)
+	if err != nil {
+		opRet = -1
+		opError = fmt.Sprintf("Could not able to write etcd environment variable. Aborting")
+		log.WithField("error", err.Error()).Error("Could not able to write etcd environment variable. Aborting")
+	}
+
+	// Restarting etcd daemon
+	etcdCmd, err := etcdmgmt.ReStartEtcd()
```
Comment (atinmu) [inline rpc/services/peer-rpc-svc.go]: s/etcdCmd/etcdCtx

File: etcdmgmt/etcd-mgmt.go (hunk 16 above)
Comment (gaurav36) [inline etcdmgmt/etcd-mgmt.go]: SpawnETCD can be use in place of "StartETCD" function. In this function we are not starting/spawning etcd.

File: etcdmgmt/etcd-mgmt.go (hunk 19 above)
Comment (gaurav36) [inline etcdmgmt/etcd-mgmt.go]: this loop needed for restoring etcd environment variable during restarting of etcd/glusterd. this loop reading etcd environment variable from the etcdConfFile and spliting   etcd environment key and value. left side of first occurrence of = will be etcdenv key and right side will be value. split function give you array of key/value which are separated by   "=" so at right side we need to join all that key/value, that is done by inner for loop.

File: rpc/services/peer-rpc-svc.go (hunk 32 above)
Comment (gaurav36) [inline rpc/services/peer-rpc-svc.go]: done

File: rpc/services/peer-rpc-svc.go (hunk 27 above)
Comment (gaurav36) [inline rpc/services/peer-rpc-svc.go]: done

File: etcdmgmt/etcd-mgmt.go (hunk 24 above)
Comment (gaurav36) [inline etcdmgmt/etcd-mgmt.go]: done

File: etcdmgmt/etcd-mgmt.go (hunk 18 above)
Comment (gaurav36) [inline etcdmgmt/etcd-mgmt.go]: done

File: commands/peers/addpeer.go [hunk 33]
```diff
@@ -43,11 +50,68 @@ func addPeerHandler(w http.ResponseWriter, r *http.Request) {
 		rest.SendHTTPError(w, http.StatusInternalServerError, *rsp.OpError)
 		return
 	}
+
+	c := context.EtcdClient
+	// Add member to etcd server
+	mAPI := etcdclient.NewMembersAPI(c)
+	member, e := mAPI.Add(etcdcontext.Background(), "http://"+p.Name+":2380")
+	if e != nil {
+		log.WithField("err", e).Error("Failed to add member into etcd cluster")
```
Comment (atinmu) [inline commands/peers/addpeer.go]: I still don't see the member information captured in the log.

File: etcdmgmt/etcd-mgmt.go [hunk 34]
```diff
@@ -160,3 +151,98 @@ func isETCDStartNeeded() (bool, int) {
...
+			"-listen-peer-urls", listenPeerUrls,
+			"-initial-advertise-peer-urls", initialAdvPeerUrls}
+
+		log.Info("Sstarting etcd daemon")
+		return StartETCD(args)
+	}
+	return nil, err
+}
+
+// Starting default etcd by considering single server node
```
Comment (atinmu) [inline etcdmgmt/etcd-mgmt.go]: A comment should in the form of "What function name and what it does". Take it as a global comment

File: rpc/services/peer-rpc-svc.go (hunk 26 above)
Comment (atinmu) [inline rpc/services/peer-rpc-svc.go]: /var/lib/glusterd/etcd/ would be a better place holder, isn't it?

File: context/context.go [hunk 35]
```diff
@@ -43,6 +46,21 @@ func initOpVersion() {
 	OpVersion = MaxOpVersion
 }
 
+func initEtcdClient() error {
+	Hostname, err := os.Hostname()
```
Comment (atinmu) [inline context/context.go]: this should be done at doInit()

Comment (atinmu) [inline context/context.go]: Hostname would be blank when etcd start is triggered as we initialize context post (re)starting etcd

File: etcdmgmt/etcd-mgmt.go [hunk 36]
```diff
@@ -1,24 +1,38 @@
...
 	"github.com/gluster/glusterd2/utils"
 
 	log "github.com/Sirupsen/logrus"
 )
 
 // ExecName is to indicate the executable name, useful for mocking in tests
-var ExecName = "etcd"
+var (
+	ExecName           = "etcd"
+	listenClientUrls   = "http://" + context.Hostname + ":2379"
```
Comment (atinmu) [inline etcdmgmt/etcd-mgmt.go]: Instead of context.Hostname I suggest we should use utils.GetLocalIP() where the function should be defined as following in utils.go:

+func GetLocalIP() string {
+       addrs, err := net.InterfaceAddrs()
+       if err != nil {
+               log.Error(err)
+               return ""
+       }
+       log.Info(addrs)
+       for _, address := range addrs {
+               // check the address type and if it is not a loopback the display it
+               if ipnet, ok := address.(*net.IPNet); ok && !ipnet.IP.IsLoopback() {
+                       if ipnet.IP.To4() != nil {
+                               return ipnet.IP.String()
+                       }
+               }
+       }
+
+       return ""
+}

File: context/context.go (hunk 35 above)
Comment (gaurav36) [inline context/context.go]: It should not be done in doInit because i am using context.Hostname before doInit() call.

File: etcdmgmt/etcd-mgmt.go (hunk 34 above)
Comment (gaurav36) [inline etcdmgmt/etcd-mgmt.go]: I didn't know about this format and saw lot of place in current existing code where nothing there as a comment in current existing code. Any way i will do this change.

File: commands/peers/addpeer.go (hunk 33 above)
Comment (gaurav36) [inline commands/peers/addpeer.go]: that one log left. will do in next PR

File: context/context.go (hunk 35 above)
Comment (atinmu) [inline context/context.go]: You can't use it without the initialization is the conclusion we have right? And more over if we are going with the IP based approach what I suggested in another comment, this is no more a valid case.

File: context/context.go [hunk 37]
```diff
@@ -63,6 +85,11 @@ func Init() {
 	initOnce.Do(doInit)
 }
 
+func GetEtcdMemberAPI() etcdclient.MembersAPI {
+	var c etcdclient.Client
```
Comment (atinmu) [inline context/context.go]: Comment missing for the function

File: etcdmgmt/etcd-mgmt.go [hunk 38]
```diff
@@ -160,3 +151,114 @@ func isETCDStartNeeded() (bool, int) {
...
+
+		log.Info("Sstarting etcd daemon")
+		return StartETCD(args)
+	}
+	return nil, err
+}
+
+//StartStandAloneETCD() will Start default etcd by considering single server node
+func StartStandAloneETCD() (*os.Process, error) {
+	args := []string{"-listen-client-urls", listenClientUrls,
```
Comment (atinmu) [inline etcdmgmt/etcd-mgmt.go]: You'd need to call initETCDArgVar()

File: etcdmgmt/etcd-mgmt.go [hunk 39]
```diff
@@ -160,3 +151,114 @@ func isETCDStartNeeded() (bool, int) {
...
+	if err != nil {
+		log.Error("Could not able to kill etcd daemon")
+		return err
+	}
+	return nil
+}
+
+// ReStartETCD() will restart etcd process
+func ReStartETCD() (*os.Process, error) {
+	// Stop etcd process
```
Comment (atinmu) [inline etcdmgmt/etcd-mgmt.go]: Same as above.

File: etcdmgmt/etcd-mgmt.go (hunk 38 above)
Comment (gaurav36) [inline etcdmgmt/etcd-mgmt.go]: its not good to call every time getLocalIP(). we just need to call initETCDArgVar at the time of starting glusterd. it will save call request.

Comment (atinmu) [inline etcdmgmt/etcd-mgmt.go]: And this is why the regression has failed?

Comment (gaurav36) [inline etcdmgmt/etcd-mgmt.go]: due to some missing part in test case.
